posts/2023-09-30-keycloak-with-openliberty.adoc

```diff
@@ -0,0 +1,126 @@
+        clientId = "sample-openliberty-keycloak",
+        clientSecret = "x4fRVAhk49TKDqVlzIt4q9oh8DSWfePt",
+        redirectToOriginalResource = true,
+        logout = @LogoutDefinition(notifyProvider = true))
+----
+
+In this example, the `providerURI` is the discovery endpoint for the `openliberty` realm, the `clientId` is the client ID of the `sample-openliberty-keycloak` client, and the `clientSecret` is the secret belonging to the `sample-openliberty-keycloak` client. By default, the redirect URI is set to `http://localhost:9090/Callback` and `redirectToOriginalResource` is set to `true` to redirect users from the redirect URI back to the originally requested resource. Lastly, `notifyProvider` in the `@LogoutDefinition` is set to `true` to also log the user out of Keycloak when a logout occurs in your Open Liberty application.
+
+Now, with this annotation set up, your REST endpoints can be protected using the `@RolesAllowed` annotation which will trigger the authentication flow when a user tries to access the endpoint.
+After authentication, the user's access token can be obtained using the `OpenIdContext`.
+
+This example shows a JAX-RS resource which contains a `/username` endpoint which is only accessible by users with the `admin` role and a `/os` endpoint which is accessible by both users with the `admin` role and users with the `user` role.
+
+[source, java]
+----
+@ApplicationScoped
+@Path("/system/properties")
+public class SystemResource {
+
+    @Inject
+    @RestClient
+    private SystemService systemService;
+
+    @Inject
+    private OpenIdContext openIdContext;
+
+    @GET
+    @Path("/username")
+    @RolesAllowed({ "admin" })
+    public String getUsername() {
+        return systemService.getUsername(openIdContext.getAccessToken().getToken());
+    }
+
+    @GET
+    @Path("/os")
+    @RolesAllowed({ "admin", "user" })
+    public String getOS() {
+        return systemService.getOS(openIdContext.getAccessToken().getToken());
+    }
+    
+}
+----
+
+Once the requests to these endpoints are authenticated and authorized, the endpoint can now use the access token of the authenticated user.
+In the above example, the access token is used as a bearer token to make a request to another protected resource by including it in the request header in the format `Authorization: Bearer <access-token>`.
+The next section will demonstrate how this bearer token can be consumed by an Open Liberty application using MicroProfile JWT to protect its resources.
+
+== Consuming an access token from Keycloak using MicroProfile JWT
+
+link:https://openliberty.io/docs/latest/reference/feature/mpJwt-2.1.html[MicroProfile JWT] can easily be used to consume access tokens sent as bearer tokens.
+
+This example shows the link:https://openliberty.io/docs/latest/microprofile-config-properties.html#jwt[MicroProfile Config properties] required to validate an access token issued by the `openliberty` realm in Keycloak.
```

Comment:
Feels like MP config came out of the blue in this part. Maybe add in why it's being used? Or into it more.